src/Surfnet/StepupGateway/SecondFactorOnlyBundle/Controller/SecondFactorOnlyController.php

```diff
@@ -132,33 +133,33 @@ public function respondAction(Request $request)
         // Reset state
         $this->getSecondFactorRespondService()->resetRespondState($responseContext);
 
-        // Handle SAML response
-        $httpResponse =  $responseRendering->renderResponse($responseContext, $response, $request);
-
-        $ssoCookieService = $this->get('gateway.service.sso_2fa_cookie');
-        $ssoCookieService->handleSsoOn2faCookieStorage($responseContext, $request, $httpResponse);
-
-        // We can now forget the selected second factor.
-        $responseContext->finalizeAuthentication();
-
-        // Check if ADFS response
+        // Check if ADFS response, if it is, we use the ADFS ACS twig template
         $adfsParameters = $this->getSecondFactorAdfsService()->handleAdfsResponse($logger, $responseContext);
-
         if (!is_null($adfsParameters)) {
             // Handle Adfs response
             $responseRendering = $this->get('second_factor_only.response_rendering');
             $xmlResponse = $responseRendering->getResponseAsXML($response);
 
-            return $this->render(
+            $httpResponse = $this->render(
                 '@SurfnetStepupGatewaySecondFactorOnly/adfs/consume_assertion.html.twig',
                 [
                     'acu' => $responseContext->getDestinationForAdfs(),
                     'samlResponse' => $xmlResponse,
                     'adfs' => $adfsParameters,
                 ]
             );
+        } else {
```

Comment:
The else expression ensures the render method is run but once. Another way of fixing this would have been to move this logic to a separate service. Or by decoupling the two code paths even further (dedicated ADFS routes).

See the TD ticket linked in the story to join the discussion in how to fix this better.